14_clusters/simple_torch_cluster.py

```diff
@@ -0,0 +1,134 @@
+# # Simple PyTorch cluster
+
+# This example shows how you can perform distributed computation with PyTorch.
+# It is a kind of 'hello world' example for distributed ML training: setting up a cluster
+# and executing a broadcast operation to share a single tensor.
+
+# ## Basic setup: Imports, dependencies, and a script
+
+# Let's get the imports out of the way first.
+# We need to import `modal.experimental` to use this feature, since it's still under development.
+# Let us know if you run into any issues!
+
+import os
+from pathlib import Path
+
+import modal
+import modal.experimental
+
+# Communicating between nodes in a cluster requires communication libraries.
+# We'll use `torch`, so we add it to our container's [Image](https://modal.com/docs/guide/images) here.
+
+image = modal.Image.debian_slim(python_version="3.12").pip_install(
+    "torch~=2.5.1", "numpy~=2.2.1"
+)
+
+# The approach we're going to take is to use a Modal [Function](https://modal.com/docs/reference/modal.Function)
+# to launch the underlying script we want to distribute over the cluster nodes.
+# The script is located in another file in the same directory
+# of [our examples repo](https://github.com/modal-labs/modal-examples/).
+# In order to use it in our remote Modal Function,
+# we need to duplicate it remotely, which we do with `add_local_file`.
+
+this_directory = Path(__file__).parent
+
+image = image.add_local_file(
+    this_directory / "simple_torch_cluster_script.py",
+    remote_path="/root/script.py",
+)
+
+app = modal.App("example-simple-torch-cluster", image=image)
+
+# ## Configuring a test cluster
+
+# Some basic configuration allows for demoing either a CPU-only cluster or a GPU-enabled cluster
+# with one GPU per container.
+
+# First, we set the size of the cluster in containers/nodes. This can be between 1 and 8.
+# This is part of our Modal configuration, since Modal is responsible for spinning up our cluster.
+
+n_nodes = 4
+
+# Next, we set the number of processes we run per node.
+# The usual practice is to run one process per GPU,
+# so we set those two values to be equal.
+# Note that `N_GPU` is Modal configuration ("how many GPUs should we spin up for you?")
+# while `nproc_per_node` is `torch.distributed` configuration ("how many processes should we spawn for you?").
+
+n_proc_per_node = N_GPU = 1
```

Comment:
what if we default to 8? after RDMA, we may want to require all GPU clusters to always have 8 GPUs.

Reply:
we would prefer to set it to 8, but then it won't work for free-tier customers, who are limited to 10 GPUs